healthmanager/blockchain_health_manager.go

```diff
@@ -0,0 +1,230 @@
+package healthmanager
+
+import (
+	"context"
+	"fmt"
+	"github.com/status-im/status-go/healthmanager/aggregator"
+	"github.com/status-im/status-go/healthmanager/rpcstatus"
+	"sync"
+)
+
+// BlockchainFullStatus contains the full status of the blockchain, including provider statuses.
+type BlockchainFullStatus struct {
+	Status                    rpcstatus.ProviderStatus                       `json:"status"`
+	StatusPerChainPerProvider map[uint64]map[string]rpcstatus.ProviderStatus `json:"statusPerChainPerProvider"`
+}
+
+// BlockchainStatus contains the status of the blockchain
+type BlockchainStatus struct {
+	Status         rpcstatus.ProviderStatus            `json:"status"`
+	StatusPerChain map[uint64]rpcstatus.ProviderStatus `json:"statusPerChain"`
+}
+
+// BlockchainHealthManager manages the state of all providers and aggregates their statuses.
+type BlockchainHealthManager struct {
+	mu          sync.RWMutex
+	aggregator  *aggregator.Aggregator
+	subscribers []chan struct{}
+
+	providers   map[uint64]*ProvidersHealthManager
+	cancelFuncs map[uint64]context.CancelFunc // Map chainID to cancel functions
+	lastStatus  BlockchainStatus
+	wg          sync.WaitGroup
+}
+
+// NewBlockchainHealthManager creates a new instance of BlockchainHealthManager.
+func NewBlockchainHealthManager() *BlockchainHealthManager {
+	agg := aggregator.NewAggregator("blockchain")
+	return &BlockchainHealthManager{
+		aggregator:  agg,
+		providers:   make(map[uint64]*ProvidersHealthManager),
+		cancelFuncs: make(map[uint64]context.CancelFunc),
+	}
+}
+
+// RegisterProvidersHealthManager registers the provider health manager.
+// It prevents registering the same provider twice for the same chain.
+func (b *BlockchainHealthManager) RegisterProvidersHealthManager(ctx context.Context, phm *ProvidersHealthManager) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	// Check if the provider for the given chainID is already registered
+	if _, exists := b.providers[phm.ChainID()]; exists {
+		// Log a warning or return an error to indicate that the provider is already registered
+		return fmt.Errorf("provider for chainID %d is already registered", phm.ChainID())
+	}
+
+	// Proceed with the registration
+	b.providers[phm.ChainID()] = phm
+
+	// Create a new context for the provider
+	providerCtx, cancel := context.WithCancel(ctx)
+	b.cancelFuncs[phm.ChainID()] = cancel
+
+	statusCh := phm.Subscribe()
+	b.wg.Add(1)
+	go func(phm *ProvidersHealthManager, statusCh chan struct{}, providerCtx context.Context) {
+		defer func() {
+			b.wg.Done()
```

Comment:
Isn't it safer to call the line below first and then this one?

Reply:
Yeah, true